java/vector/src/main/java/org/apache/arrow/vector/BitVector.java

```diff
@@ -18,342 +18,469 @@
-  int valueCount;
-  private int allocationSizeInBytes = getSizeFromCount(INITIAL_VALUE_ALLOCATION);
-  private int allocationMonitor = 0;
+public class BitVector extends BaseFixedWidthVector {
+  private final FieldReader reader;
 
+  /**
+   * Instantiate a BitVector. This doesn't allocate any memory for
+   * the data in vector.
+   *
+   * @param name      name of the vector
+   * @param allocator allocator for memory management.
+   */
   public BitVector(String name, BufferAllocator allocator) {
-    super(name, allocator);
+    this(name, FieldType.nullable(Types.MinorType.BIT.getType()),
+            allocator);
   }
 
-  @Override
-  public void load(ArrowFieldNode fieldNode, ArrowBuf data) {
-    // When the vector is all nulls or all defined, the content of the buffer can be omitted
-    if (data.readableBytes() == 0 && fieldNode.getLength() != 0) {
-      int count = fieldNode.getLength();
-      allocateNew(count);
-      int n = getSizeFromCount(count);
-      if (fieldNode.getNullCount() == 0) {
-        // all defined
-        // create an all 1s buffer
-        // set full bytes
-        int fullBytesCount = count / 8;
-        for (int i = 0; i < fullBytesCount; ++i) {
-          this.data.setByte(i, 0xFF);
-        }
-        int remainder = count % 8;
-        // set remaining bits
-        if (remainder > 0) {
-          byte bitMask = (byte) (0xFFL >>> ((8 - remainder) & 7));
-          this.data.setByte(fullBytesCount, bitMask);
-        }
-      } else if (fieldNode.getNullCount() == fieldNode.getLength()) {
-        // all null
-        // create an all 0s buffer
-        zeroVector();
-      } else {
-        throw new IllegalArgumentException("The buffer can be empty only if there's no data or it's all null or all defined");
-      }
-      this.data.writerIndex(n);
-    } else {
-      super.load(fieldNode, data);
-    }
-    this.valueCount = fieldNode.getLength();
+  /**
+   * Instantiate a BitVector. This doesn't allocate any memory for
+   * the data in vector.
+   *
+   * @param name      name of the vector
+   * @param fieldType type of Field materialized by this vector
+   * @param allocator allocator for memory management.
+   */
+  public BitVector(String name, FieldType fieldType, BufferAllocator allocator) {
+    super(name, allocator, fieldType, (byte) 0);
+    reader = new BitReaderImpl(BitVector.this);
   }
 
+  /**
+   * Get a reader that supports reading values from this vector
+   *
+   * @return Field Reader for this vector
+   */
   @Override
-  public Field getField() {
-    throw new UnsupportedOperationException("internal vector");
+  public FieldReader getReader() {
+    return reader;
   }
 
+  /**
+   * Get minor type for this vector. The vector holds values belonging
+   * to a particular type.
+   *
+   * @return {@link org.apache.arrow.vector.types.Types.MinorType}
+   */
   @Override
-  public MinorType getMinorType() {
-    return MinorType.BIT;
+  public Types.MinorType getMinorType() {
```

Comment:
Why not "MinorType"?

Reply:
Hmm, I think this must have been auto-corrected, let me check on it

Reply:
It looks like the other vector classes also use `Types.MinorType`.  I agree, it would be better to just say `MinorType`

Reply:
@BryanCutler I think this is missed. Can you please fix this?